internal/daemon/daemon.go

```diff
@@ -154,6 +149,37 @@ func (d *Daemon) init(extendedEndpoints []rest.Endpoint, schemaExtensions map[in
 	return nil
 }
 
+func (d *Daemon) applyHooks(hooks *config.Hooks) {
+	// Apply the default empty hooks if no hooks are given.
+	defaultHook := func(s *state.State) error { return nil }
+
+	if hooks == nil {
+		d.hooks = config.Hooks{}
+	} else {
+		d.hooks = *hooks
+	}
+
+	if d.hooks.OnBootstrapHook == nil {
+		d.hooks.OnBootstrapHook = defaultHook
+	}
+
+	if d.hooks.OnJoinHook == nil {
+		d.hooks.OnJoinHook = defaultHook
+	}
+
+	if d.hooks.OnStartHook == nil {
+		d.hooks.OnStartHook = defaultHook
+	}
+
+	if d.hooks.OnHeartbeatHook == nil {
+		d.hooks.OnHeartbeatHook = defaultHook
+	}
+
+	if d.hooks.OnRemoveHook == nil {
+		d.hooks.OnRemoveHook = defaultHook
```

Comment:
What does the defaultHook do? Is it required ?

Reply:
It's just an empty function, so we don't have to constantly check if a hook is nil before trying to run it.

Reply:
Ah ok maybe call it noopHook